habitat/hooks/run

```diff
@@ -2,59 +2,55 @@
+TLS_ADDRESS="{{ sys.ip }}:{{ cfg.tpp.device.tls_port }}"
+{{/unless}}
+{{#if cfg.tpp.device.tls_address ~}}
+TLS_ADDRESS="{{ cfg.tpp.device.tls_address }}:{{ cfg.tpp.device.tls_port }}"
+{{/if ~}}
+
+# Set certificate file names
 CERT_FILE="${COMMON_NAME}.cert"
 KEY_FILE="${COMMON_NAME}.key"
 CHAIN_FILE="${COMMON_NAME}.chain"
+ID_FILE="${COMMON_NAME}.id"
 
 # Set cert path
 CERT_PATH="{{pkg.svc_data_path}}/${CERT_FILE}"
 KEY_PATH="{{pkg.svc_data_path}}/${KEY_FILE}"
 CHAIN_PATH="{{pkg.svc_data_path}}/${CHAIN_FILE}"
-CERT_ID="\\\\VED\\\\Policy\\\\${CERTIFICATE_ZONE}\\\\${COMMON_NAME}"
-
-HOSTNAME="{{sys.hostname}}"
-
-# Set DN paths for device and app creation
-DEVICE_CREATION_DN="\\\\VED\\\\Policy\\\\Devices And Applications\\\\External\\\\${HOSTNAME}"
-APP_CREATION_DN="\\\\VED\\\\Policy\\\\Devices And Applications\\\\External\\\\${HOSTNAME}\\\\${COMMON_NAME}-{{sys.ip}}{{sys.hostname}}"
-
-RUN_ONCE=true
+ID_PATH="{{pkg.svc_data_path}}/${ID_FILE}"
 
 # Establish a CA
 SSL_CERT_FILE="{{pkgPathFor "core/cacerts"}}/ssl/certs/cacert.pem"
 export SSL_CERT_FILE
 
+# https://support.venafi.com/hc/en-us/articles/217991528-Introducing-VCert-API-Abstraction-for-DevOps
 function enroll_new_cert() {
-    # Enroll a new SSL cert
-    exec vcert enroll -no-prompt \
-        -tpp-url ${TPP_URL} \
-        -tpp-user ${TPP_USER} \
-        -tpp-password ${TPP_PASSWORD} \
-        -cert-file ${CERT_PATH} \
-        -key-file ${KEY_PATH} \
-        -chain-file ${CHAIN_PATH} \
-        -cn ${COMMON_NAME} \
-        -z ${CERTIFICATE_ZONE}
+    exec vcert enroll -no-prompt {{#if cfg.cloud.auth.apikey ~}} -k ${CLOUD_APIKEY} {{/if ~}}{{#if cfg.tpp.auth.url ~}} -tpp-user ${TPP_USER} -tpp-password ${TPP_PASSWORD} -u ${TPP_URL} {{/if ~}} -z "{{ cfg.zone }}" -cert-file ${CERT_PATH} -key-file ${KEY_PATH} -chain-file ${CHAIN_PATH} -cn ${COMMON_NAME} -pickup-id-file ${ID_PATH} {{#if cfg.tpp.device.register ~}}-instance ${INSTANCE} -replace-instance -app-info ${APP_INFO} -tls-address ${TLS_ADDRESS}{{/if ~}}
+    
+    # TODO: This 'echo' command is filler because the "if" helper
+    # On the previous line pulls the curly brace that closes the
+    # function to the previous line (invalid shell syntax). 
+    # Remove when this is resolved.
+    echo "Proceeding to renewal check."
 }
 
 function renew_cert() {
-    # Renew the SSL cert
-    exec vcert renew -no-prompt \
-        -tpp-url ${TPP_URL} \
-        -tpp-user ${TPP_USER} \
-        -tpp-password ${TPP_PASSWORD} \
-        -cert-file ${CERT_PATH} \
-        -key-file ${KEY_PATH} \
-        -chain-file ${CHAIN_PATH} \
-        -id ${CERT_ID} \
-        -z ${CERTIFICATE_ZONE}
+    exec vcert renew -no-prompt {{#if cfg.cloud.auth.apikey ~}} -k ${CLOUD_APIKEY} {{/if ~}} {{#if cfg.tpp.auth.url ~}} -tpp-url ${TPP_URL} -tpp-user ${TPP_USER} -tpp-password ${TPP_PASSWORD} {{/if ~}} -cert-file ${CERT_PATH} -key-file ${KEY_PATH} -chain-file ${CHAIN_PATH} -id file:${ID_PATH} -z "{{ cfg.zone }}"
```

Comment:
Here's a cool think I discovered while writing a new hab package recently. 'exec' executes the command and replaces the shell with the new process that you executed. This leads to the issue I experienced before where the Habitat run hook would quit and restart as soon as the vcert operation would quit. Just as an aside, if you don't want the vcert process to terminate the run hook as soon as it's done you can do something like:

`CMD="exec vcert renew -no-prompt {{#if cfg.cloud.auth.apikey ~}} -k ${CLOUD_APIKEY} {{/if ~}} {{#if cfg.tpp.auth.url ~}} -tpp-url ${TPP_URL} -tpp-user ${TPP_USER} -tpp-password ${TPP_PASSWORD} {{/if ~}} -cert-file ${CERT_PATH} -key-file ${KEY_PATH} -chain-file ${CHAIN_PATH} -id file:${ID_PATH} -z \"{{ cfg.zone }\}""
${CMD}`

Reply:
Totally! Usually this is done so that habitat manages the proc, in this case the `vcert` process. If the intended result is for habitat to manage the run hook, rather than `vcert`, then I would suggest removing the `exec`.

Judging by the need to sleep in the run hook, I think it makes sense to opt-out of the exec.